### PR TITLE
[NET-8946 NET-8947 NET-8948] security: bump go, x/net and envoy versions

### PR DESCRIPTION
**Pull request by dduzgun-security**
Description
Upgrade to use Go 1.21.9. This resolves CVEs
[CVE-2023-45288](https://nvd.nist.gov/vuln/detail/CVE-2023-45288) (http2).

Upgrade to support Envoy 1.28.2. This resolves CVEs
[CVE-2024-27919](https://nvd.nist.gov/vuln/detail/CVE-2024-27919) (http2).

Upgrade to use golang.org/x/net v0.24.0. This resolves CVEs
[CVE-2023-45288](https://nvd.nist.gov/vuln/detail/CVE-2023-45288) (x/net).

Checklist
 Tests added
 [CHANGELOG entry added](https://github.com/hashicorp/consul-k8s/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -27,7 +27,7 @@ RUN apt-get update && apt install -y libcap2-bin
 RUN setcap CAP_NET_BIND_SERVICE=+ep /usr/local/bin/envoy
 RUN setcap CAP_NET_BIND_SERVICE=+ep /usr/local/bin/$BIN_NAME
 
-FROM hashicorp/envoy-fips:1.28.1-fips1402 as envoy-fips-binary
+FROM hashicorp/envoy-fips:1.28.2-fips1402 as envoy-fips-binary
```
Comment (dduzgun-security) [inline Dockerfile:30]: I think the tests are failing due to the `hashicorp/envoy-fips:1.28.2-fips1402` container image that doesn't yet exist. Not sure on how to publish a new one, some guidance would be helpful 👀

Comment (zalimeni) [reply Dockerfile:30]: That's correct. There's an effort right now to reconfigure ownership of that publishing going forward; I'll drop a question in Slack to see if we can get a hand. For now, @dduzgun-security I'd omit this and put in a TODO that it needs to be bumped before the next release (that's what we did last time)

**Review by zalimeni — APPROVED** [review]
Thank you @dduzgun-security ! As w/ https://github.com/hashicorp/consul-k8s/pull/3893, we'll want to backport this through 1.1.x.

File: Dockerfile (hunk 1 above)
Comment (dduzgun-security) [reply Dockerfile]: Thanks for initiating the Slack conversation, I have reverted the change: https://github.com/hashicorp/consul-dataplane/pull/474/commits/90790c1267c76b964c669dabcf05d20db449bd6a and added ticket [NET-9002] to revisit it before the next release.

**Pull request merged**